### PR TITLE
[SPARK-39091][SQL] Updating specific SQL Expression traits that don't compose when multiple are extended due to nodePatterns being final.

### PR DESCRIPTION
**Pull request by neilagupta**
When creating classes that extend multiple `Expression` traits, they have an inheritance conflict due to `nodePatterns` being different values. `nodePatterns` is also final, which makes the inheritance conflict unavoidable.

This error comes in after trying to override the field - `value nodePatterns in trait TimeZoneAwareExpression of type Seq[org.apache.spark.sql.catalyst.trees.TreePattern.TreePattern] cannot override final member.
`

### What changes were proposed in this pull request?

For now kept the traits changed to a minimum - removed `nodePatterns` final classifier from `TimeZoneAwareExpression`, `HigherOrderFunctions`, and `Generator`. I have propagated this change to more children classes of Expression, but this change can be expanded (or unexpanded) to other traits if needed.

### Why are the changes needed?

Enables multiple trait inheritance for certain `Expression` types. Hand-picked a few of them to have their `nodePatterns` attribute not final.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a test that creates a dummy class that inherits all three of `TimeZoneAwareExpression`, `HigherOrderFunctions`, and `Generator` - the test is only to see if this can compile now. Since it compiles and runs, this issue is resolved for those three traits at minimum.

### CODE REVIEW

**Comment by HuwCampbell** [comment]
This change would be valuable for me. I imagine that if manually overidden, one would have to specify all bit patterns used in order to get all appropriate rules applied.

```scala
Seq(TIME_ZONE_AWARE_EXPRESSION, HIGHER_ORDER_FUNCTION, GENERATOR)
```

is that correct?

**Comment by neilagupta** [comment]
Hi @HuwCampbell that should be correct, I will update this PR with more modified subclasses of Expression to have their `nodePatterns` fields not final. Would that be in your interest?

**Comment by HuwCampbell** [comment]
Yes indeed, this would unblock me from upgrading to Spark 3.2 once there's a release.

**Comment by neilagupta** [comment]
@AmplabJenkins  any chance I could get someone with write access to review this?

**Comment by hvanhovell** [comment]
@sigmod wanna take a look?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HigherOrderFunctionsSuite.scala [hunk 1]
```diff
@@ -839,6 +841,52 @@ class HigherOrderFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper
...
+      override def bind(f: (Expression,
+        Seq[(DataType, Boolean)]) => LambdaFunction): HigherOrderFunction = null
+
+      override def timeZoneId: Option[String] = None
+
+      override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression = null
+    }
+
+    /* Should not get the error - value nodePatterns
+     ... cannot override final member, or conflict in nodePatterns between two types
```
[the remark this answers is not in the document]
Comment (HuwCampbell) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HigherOrderFunctionsSuite.scala:886]: Yeah I think you're pretty much dead on here, the correctness of the optimiser will be up to the user to enforce; as they will need to ensure that all node patterns are accounted for.

But, and it's a big one, it makes it possible to compose these interfaces, even with the extra challenges... which it's just _not_ possible to do as of Spark 2.2.

**Pull request closed without merging**